toolchains/tool_map.bzl

```diff
@@ -0,0 +1,125 @@
+# Copyright 2025 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of tool_map."""
+
+load(
+    "//toolchains/private:collect.bzl",
+    "collect_provider",
+    "collect_tools",
+)
+load(
+    ":toolchain_info.bzl",
+    "ActionTypeSetInfo",
+    "ToolMapInfo",
+)
+
+def _tool_map_impl(ctx):
+    tools = collect_tools(ctx, ctx.attr.tools)
+    action_sets = collect_provider(ctx.attr.actions, ActionTypeSetInfo)
+
+    action_to_tool = {}
+    action_to_as = {}
+    for i in range(len(action_sets)):
+        action_set = action_sets[i]
+        tool = tools[ctx.attr.tool_index_for_action[i]]
+
+        for action in action_set.actions.to_list():
+            if action in action_to_as:
+                fail("The action %s appears multiple times in your tool_map (as %s and %s)" % (action.label, action_set.label, action_to_as[action].label))
```

Comment:
No need to do this now, but if you pass labels as individual parameters to `fail`, they are prettyprinted.